### ISSUE TITLE
External load balancer

### ISSUE THREAD

**Issue by fabriziopandini** [kubernetes-sigs/kind#134]
Multi node scenario might require an external load balancer in front of control-plane instances (if more than one should be deployed in a cluster)

A possible solution is to use an HAProxy container to be created in addition to the cluster nodes, but other solutions could be considered as well.

The HAProxy solution, even if simple, introduces a "dependency" from a component not currently included in the node-image, and this should be considered as well.

opinions?
/assign
/

**Comment by neolit123** [comment]
> The HAProxy solution, even if simple, introduces a "dependency" from a component not currently included in the node-image, and this should be considered as well.

so if HAProxy is installed, but disabled by default on the default node image and we have control of what we can execute on such a node we can set this node as an external LB.

one problem here is that the node-image is pretty big and i don't think it should use it for just this purpose, but for testing it seems like an option to consider as well.

i think we should start estimating RAM usage for common setups too.

i've commented here on the related topic:
https://github.com/kubernetes-sigs/kind/issues/135#issuecomment-442670776

**Comment by BenTheElder** [comment]
how big is a haproxy install, exactly?

I've requested that we try to keep things versioned within that image. This means that eg our first release has two artifacts:
- `kind` CLI binaries
- a pinned image used by those binaries, which we can never change because we pin it by the sha.

for the default usage, absolutely everything is stable in this case (well except weave, but that's a bug not a pattern to follow)

If it's not too big, the image downloads are compressed anyhow and the containers should be COW, so it's probably ok to just add it to the base image.

**Comment by neolit123** [comment]
this page tells me 1MB for the DEB of recent haproxy:
https://ubuntu.pkgs.org/18.10/ubuntu-main-amd64/haproxy_1.8.13-2build1_amd64.deb.html

> If it's not too big, the image downloads are compressed anyhow and the containers should be COW, so it's probably ok to just add it to the base image.

oh, copy on write, of course. i forgot about this magic.
i see the the CNI plugins are part of the base image so its seems like the right place for it.

**Comment by fabriziopandini** [comment]
@BenTheElder it is time to start implementing external load balancer

If I got it right solutions discussed above are:
1. Use ha proxy image for external external load balancing nodes
2. Use kind default image, start docker in docker, start haproxy using docker in docker (eventually pre-loaded)

1 is simplest, and probably with less footprint, but 2 address the foundament requirement keeping things versioned within a unique node image.
So I'm taking two as working assumption.

Please let me know if there are concerns/other things to consider

**Comment by fabriziopandini** [comment]
@BenTheElder yes. In the draft per I'm using a kind node that simply runs ha proxy

**Comment by BenTheElder** [comment]
OK, but to clarify that is neither 1) not 2) though. It's basically 2) but drop running it in another layer of docker. (?)
